File: js/openaiChatCompletion/multimodal.ts
import "dotenv/config";
// @ts-ignore
import fs from "fs";
import { wrapOpenAI } from "langsmith/wrappers";
import { OpenAI } from "openai";
import { image } from "../resources/base64BufoApproval";
import { milwaukeeArtMuseumPDFBase64 } from "../resources/Milwaukee Art Museum PDF";
const client = wrapOpenAI(new OpenAI());

async function imageUrlExample() {
  const response = await client.chat.completions.create({
    model: "gpt-5-2025-08-07",
    messages: [
      {
        role: "user",
        content: [
          {
            type: "text",
            text: "What's in this image?",
          },
          {
            type: "image_url",
            image_url: {
              url: "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
            },
          },
        ],
      },
    ],
  });

  console.log("Image URL response:");
  console.log(response.choices[0].message.content);
}

async function imageBase64Example() {
  const response = await client.chat.completions.create({
    model: "gpt-5-2025-08-07",
    messages: [
      {
        role: "user",
        content: [
          {
            type: "text",
            text: "Describe this image in detail.",
          },
          {
            type: "image_url",
            image_url: {
              url: image,
            },
          },
        ],
      },
    ],
  });

  console.log("\nImage base64 response:");
  console.log(response.choices[0].message.content);
}

async function pdfUploadExample() {
  const file = await client.files.create({
    file: fs.createReadStream("./resources/Milwaukee Art Museum Summary.pdf"),
    purpose: "user_data",
  });
  const response = await client.chat.completions.create({
    model: "gpt-5-2025-08-07",
    messages: [
      {
        role: "user",
        content: [
          {
            type: "text",
            text: "Please summarize this PDF document.",
          },
          {
            type: "file",
            file: {
              file_id: file.id,
            },
          },
        ],
      },
    ],
  });

  console.log("\nPDF response:");
  console.log(response.choices[0].message.content);
}

async function pdfExampleBase64() {
  const response = await client.chat.completions.create({
    model: "gpt-5-2025-08-07",
    messages: [
      {
        role: "user",
        content: [
          {
            type: "text",
            text: "Please summarize this PDF document.",
          },
          {
            type: "file",
            file: {
              filename: "milwaukeeArtMuseum.pdf",
              file_data: `data:application/pdf;base64,${milwaukeeArtMuseumPDFBase64}`,
            },
          },
        ],
      },
    ],
  });

  console.log("\nPDF response:");
  console.log(response.choices[0].message.content);
}

async function audioInputExample() {
  // Fetch an audio file and convert it to a base64 string
  const url = "https://cdn.openai.com/API/docs/audio/alloy.wav";
  const audioResponse = await fetch(url);
  const buffer = await audioResponse.arrayBuffer();

  // Convert ArrayBuffer to base64
  const bytes = new Uint8Array(buffer);
  let binary = "";
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  const base64str = btoa(binary);

  const response = await client.chat.completions.create({
    model: "gpt-4o-audio-preview",
    modalities: ["text", "audio"],
    audio: { voice: "alloy", format: "wav" },
    messages: [
      {
        role: "user",
        content: [
          { type: "text", text: "What is in this recording?" },
          {
            type: "input_audio",
            input_audio: { data: base64str, format: "wav" },
          },
        ],
      },
    ],
    store: true,
  });

  console.log(response.choices[0]);
}

async function audioOutputExample() {
  const response = await client.chat.completions.create({
    model: "gpt-4o-audio-preview",
    modalities: ["text", "audio"],
    audio: { voice: "alloy", format: "wav" },
    messages: [{ role: "user", content: "Hello! Tell me a short joke." }],
  });

  console.log("\nAudio output response:");
  console.log("Text:", response.choices[0].message.content);
  console.log("Audio data available:", !!response.choices[0].message.audio);
}

export async function main() {
  await imageUrlExample();
  await imageBase64Example();
  await pdfUploadExample();
  await pdfExampleBase64();
  await audioInputExample();
  await audioOutputExample();
}

File: python/wrap_openai_chat/multimodal.py
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))

import openai
from helpers import fetch_image_base64, fetch_pdf_base64
from langsmith import traceable
from langsmith.wrappers import wrap_openai

client = wrap_openai(openai.Client())


@traceable(name="Chat Completion Image URL")
def image_url_example():
    """Image input using URL."""
    response = client.chat.completions.create(
        model="gpt-5-2025-08-07",
        messages=[
            {
                "role": "user",
                "content": [
                    {
                        "type": "text",
                        "text": "What's in this image?",
                    },
                    {
                        "type": "image_url",
                        "image_url": {
                            "url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
                        },
                    },
                ],
            },
        ],
    )

    print(f"Image URL response: {response.choices[0].message.content[:100]}...")
    return response


@traceable(name="Chat Completion Image Base64")
def image_base64_example():
    """Image input using base64."""
    image_url = "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg"
    image_data = fetch_image_base64(image_url)

    response = client.chat.completions.create(
        model="gpt-5-2025-08-07",
        messages=[
            {
                "role": "user",
                "content": [
                    {
                        "type": "text",
                        "text": "Describe this image in detail.",
                    },
                    {
                        "type": "image_url",
                        "image_url": {
                            "url": f"data:image/jpeg;base64,{image_data}",
                        },
                    },
                ],
            },
        ],
    )

    print(f"Image base64 response: {response.choices[0].message.content[:100]}...")
    return response


@traceable(name="Chat Completion PDF Base64")
def pdf_base64_example():
    """PDF input using base64."""
    pdf_url = "https://pdfobject.com/pdf/sample.pdf"
    pdf_data = fetch_pdf_base64(pdf_url)

    response = client.chat.completions.create(
        model="gpt-5-2025-08-07",
        messages=[
            {
                "role": "user",
                "content": [
                    {
                        "type": "text",
                        "text": "Please summarize this PDF document.",
                    },
                    {
                        "type": "file",
                        "file": {
                            "filename": "sample.pdf",
                            "file_data": f"data:application/pdf;base64,{pdf_data}",
                        },
                    },
                ],
            },
        ],
    )

    print(f"PDF response: {response.choices[0].message.content[:100]}...")
    return response


def main():
    print("Running wrap_openai chat completion multimodal examples...")
    image_url_example()
    image_base64_example()
    pdf_base64_example()
    return {"multimodal": "complete"}
